core/src/main/java/org/opensearch/sql/expression/function/CollectionUDF/MVAppendFunctionImpl.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.sql.expression.function.CollectionUDF;
+
+import static org.apache.calcite.sql.type.SqlTypeUtil.createArrayType;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.calcite.adapter.enumerable.NotNullImplementor;
+import org.apache.calcite.adapter.enumerable.NullPolicy;
+import org.apache.calcite.adapter.enumerable.RexToLixTranslator;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.linq4j.tree.Expressions;
+import org.apache.calcite.linq4j.tree.Types;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.sql.SqlOperatorBinding;
+import org.apache.calcite.sql.type.SqlReturnTypeInference;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.opensearch.sql.expression.function.ImplementorUDF;
+import org.opensearch.sql.expression.function.UDFOperandMetadata;
+
+/**
+ * MVAppend function that appends all elements from arguments to create an array. Always returns an
+ * array or null for consistent type behavior.
+ */
+public class MVAppendFunctionImpl extends ImplementorUDF {
+
+  public MVAppendFunctionImpl() {
+    super(new MVAppendImplementor(), NullPolicy.ANY);
+  }
+
+  @Override
+  public SqlReturnTypeInference getReturnTypeInference() {
+    return sqlOperatorBinding -> {
+      RelDataTypeFactory typeFactory = sqlOperatorBinding.getTypeFactory();
+
+      if (sqlOperatorBinding.getOperandCount() == 0) {
+        return typeFactory.createSqlType(SqlTypeName.NULL);
+      }
+
+      RelDataType elementType = determineElementType(sqlOperatorBinding, typeFactory);
+      return createArrayType(
+          typeFactory, typeFactory.createTypeWithNullability(elementType, true), true);
+    };
+  }
+
+  @Override
+  public UDFOperandMetadata getOperandMetadata() {
+    return null;
+  }
+
+  private static RelDataType determineElementType(
+      SqlOperatorBinding sqlOperatorBinding, RelDataTypeFactory typeFactory) {
+    RelDataType mostGeneralType = null;
+    boolean hasStringType = false;
+    boolean hasNumericType = false;
+
+    for (int i = 0; i < sqlOperatorBinding.getOperandCount(); i++) {
+      RelDataType operandType = getComponentType(sqlOperatorBinding.getOperandType(i));
+
+      if (isStringType(operandType)) {
+        hasStringType = true;
+      } else if (isNumericType(operandType)) {
+        hasNumericType = true;
+      }
+
+      mostGeneralType = updateMostGeneralType(mostGeneralType, operandType, typeFactory);
+    }
+
+    if (hasStringType && hasNumericType) {
+      return typeFactory.createSqlType(SqlTypeName.VARCHAR);
+    }
```

Comment:
3 questions
1. Could explain mostGeneralType rule? 
2. if (hasStringType && hasNumericType) it should return ANY?
3. can we leverage typeFactory.leastRestrictive(List.of(...))?